### PR TITLE
fixes #3491 - API v2 rabl templates base, main, show for each controller

### PR DESCRIPTION
**Pull request by isratrade**
base.json.rabl - usually contains `:id, :name` and possibly another important attribute

main.json.rabl - extends 'base' and adds attributes, but no :child nodes
                         - index.json.rabl extends the 'main' template

show.json.rabl - extends 'main' and adds :child nodes
                          - currently, 'show' may not have anything different other than extending 'main', but this will probably change in PR 
                            fixes #3492 to add nested routes and relationships in responses.

### CODE REVIEW

**Comment by isratrade** [comment]
@tstrachota, I completed all the controllers.

**Comment by domcleal** [comment]
Where will base be used?

**Comment by isratrade** [comment]
@domcleal in base is used in child nodes of an object.  It will be more obvious in my next PR, fixes #3492

**Comment by isratrade** [comment]
@domcleal, ping?

**Comment by mbacovsky** [comment]
Is index for host not extending main or base intentionally? If I would like to add  IP, environment and model to index to match UI, where should it go? Could you please explain a bit why is host attrs divided this way?

**Comment by ohadlevy** [comment]
I tend to agree, at least the attributes that directly assigned to the host (such as ipaddress) will not impact performance and would be very useful.

**Comment by mbacovsky** [comment]
Why audits and plugins do not follow this convention?
What do you think about max line length in rabl templates or visually grouping the attributes? Some of the views are difficult to edit/check?

Other than that I like it. At first I was hesitant about the main and base separation, but the use of base in nested resources makes a lot of sense. The views look nice and consistent this way.

**Comment by isratrade** [comment]
@mbacovsky, yes I have been meaning to add the host's attributes to be more comprehensive

**Comment by mbacovsky** [comment]
@isratrade cool, thanks!

**Comment by mbacovsky** [comment]
Some more comments:
- in operating systems I'd suggest to move major and minor attrs to base. That way we can get version to associations.
- having template attr in main in config_templates causes returning the templates itself in index. Was that intentional?

File: app/views/api/v2/architectures/show.json.rabl [hunk 1]
```diff
@@ -1,3 +1,3 @@
 object @architecture
 
-attributes :name, :id, :created_at, :operatingsystem_ids, :updated_at
```
Comment (domcleal) [inline app/views/api/v2/architectures/show.json.rabl:3]: lost operating systems?  Shouldn't this be a child node?

File: app/views/api/v2/config_templates/show.json.rabl [hunk 2]
```diff
@@ -1,11 +1,11 @@
 object @config_template
 
-attributes :id, :name, :template, :snippet, :audit_comment, :template_kind_id, :template_kind_name
+extends "api/v2/config_templates/main"
 
 child :template_combinations, :object_root => false do
   extends "api/v2/template_combinations/show"
```
Comment (domcleal) [inline app/views/api/v2/config_templates/show.json.rabl]: base?

File: app/views/api/v2/hosts/show.json.rabl [hunk 3]
```diff
@@ -1,20 +1,6 @@
...
-  attributes :organization_id, :organization_name
-end
-
-if SETTINGS[:locations_enabled]
-  attributes :location_id, :location_name
-end
+extends "api/v2/hosts/main"
 
 child :host_parameters, :object_root => false do
   attributes :id, :name, :value, :priority, :is_property, :created_at, :updated_at
```
Comment (domcleal) [inline app/views/api/v2/hosts/show.json.rabl]: Should be another rabl?  And interfaces below should be "base"?

File: app/views/api/v2/interfaces/main.json.rabl [hunk 4]
```diff
@@ -0,0 +1,5 @@
+object @interface
+
+extends "api/v2/interfaces/base"
+
+attributes :ip, :mac, :username, :password, :provider, :subnet_id, :subnet_name, :domain_id, :domain_name, :created_at, :updated_at
```
Comment (domcleal) [inline app/views/api/v2/interfaces/main.json.rabl]: MAC/IP at least might be useful in base?

File: app/views/api/v2/media/main.json.rabl [hunk 5]
```diff
@@ -0,0 +1,11 @@
+object @medium
+
+extends "api/v2/media/base"
+
+attributes :path, :os_family, :created_at, :updated_at, :operatingsystem_ids
```
Comment (domcleal) [inline app/views/api/v2/media/main.json.rabl]: Should OSes be child nodes?

File: app/views/api/v2/operatingsystems/base.json.rabl [hunk 6]
```diff
@@ -0,0 +1,3 @@
+object @operatingsystem
+
+attributes :id, :name
```
Comment (domcleal) [inline app/views/api/v2/operatingsystems/base.json.rabl]: label might be good here as name isn't unique

File: app/views/api/v2/operatingsystems/show.json.rabl [hunk 7]
```diff
@@ -1,5 +1,6 @@
 object @operatingsystem
-attributes :id, :name, :major, :minor, :family, :release_name
+
+extends "api/v2/operatingsystems/main"
 
 child :media, :object_root => false do
   attributes :id, :name
```
Comment (domcleal) [inline app/views/api/v2/operatingsystems/show.json.rabl]: should use other base templates?

File: app/views/api/v2/override_values/create.json.rabl [hunk 8]
```diff
@@ -1,3 +1,3 @@
 object @override_value
 
-extends "api/v2/override_values/show"
+extends "api/v2/override_values/base"
```
Comment (domcleal) [inline app/views/api/v2/override_values/create.json.rabl]: Looks like we have a mix of base and show in the create templates, shouldn't they be show?

File: app/views/api/v2/parameters/show.json.rabl [hunk 9]
```diff
@@ -1,3 +1,3 @@
-object @parameter => :parameter
+object @parameter
 
 attributes :id, :name, :value
```
Comment (domcleal) [inline app/views/api/v2/parameters/show.json.rabl]: Should extend main?

File: app/views/api/v2/ptables/main.json.rabl [hunk 10]
```diff
@@ -0,0 +1,5 @@
+object @ptable
+
+extends "api/v2/ptables/base"
+
+attributes :layout, :os_family, :created_at, :updated_at
```
Comment (domcleal) [inline app/views/api/v2/ptables/main.json.rabl]: layout's a bit much to put in main I think, probably better in show?

File: app/views/api/v2/config_templates/main.json.rabl [hunk 11]
```diff
@@ -0,0 +1,5 @@
+object @config_template
+
+extends "api/v2/config_templates/base"
+
+attributes :template, :snippet, :audit_comment, :template_kind_id, :template_kind_name, :created_at, :updated_at
```
Comment (domcleal) [inline app/views/api/v2/config_templates/main.json.rabl]: template is probably too much for main, should it go in show?

File: app/views/api/v2/puppetclasses/show.json.rabl [hunk 12]
```diff
@@ -1,6 +1,6 @@
 object @puppetclass
 
-attributes :id, :name, :created_at, :updated_at
+extends "api/v2/puppetclasses/main"
 
 child :environments, :object_root => false do
   attributes :id, :name
```
Comment (domcleal) [inline app/views/api/v2/puppetclasses/show.json.rabl]: use base?

File: app/views/api/v2/reports/main.json.rabl [hunk 13]
```diff
@@ -0,0 +1,19 @@
+object @report
+
+extends "api/v2/reports/base"
+
+attributes :metrics, :created_at, :updated_at
```
Comment (domcleal) [inline app/views/api/v2/reports/main.json.rabl:5]: Missing host?

File: app/views/api/v2/reports/show.json.rabl [hunk 14]
```diff
@@ -1,6 +1,6 @@
 object @report
 
-attributes :id, :reported_at, :status, :metrics
+extends "api/v2/reports/main"
 
 child :logs, :object_root => false do
```
Comment (domcleal) [inline app/views/api/v2/reports/show.json.rabl:5]: Now in main

File: app/views/api/v2/smart_class_parameters/create.json.rabl [hunk 15]
```diff
@@ -1,3 +1,3 @@
-object @smart_class_parameter => :smart_class_parameter
+object @smart_class_parameter
 
-extends "api/v2/smart_class_parameters/show"
+extends "api/v2/smart_class_parameters/base"
```
Comment (domcleal) [inline app/views/api/v2/smart_class_parameters/create.json.rabl]: not show?

File: app/views/api/v2/smart_proxies/show.json.rabl [hunk 16]
```diff
@@ -1,6 +1,6 @@
 object @smart_proxy
 
-attributes :name, :id, :url, :created_at, :updated_at
+extends "api/v2/smart_proxies/main"
 
 child :features, :object_root => false do
```
Comment (domcleal) [inline app/views/api/v2/smart_proxies/show.json.rabl]: Move to main?

File: app/views/api/v2/subnets/show.json.rabl [hunk 17]
```diff
@@ -1,8 +1,6 @@
 object @subnet
 
-attributes :id, :name, :network, :mask, :priority, :vlanid,
-  :gateway, :dns_primary, :dns_secondary, :from, :to, :domain_ids,
-  :dns_id, :dhcp_id, :tftp_id, :cidr
+extends "api/v2/subnets/main"
 
 child :dhcp => :dhcp do
   attributes :id, :name, :url
```
Comment (domcleal) [inline app/views/api/v2/subnets/show.json.rabl]: smart proxy base?

File: app/views/api/v2/template_combinations/base.json.rabl [hunk 18]
```diff
@@ -0,0 +1,3 @@
+object @template_combination
+
+attributes :id, :config_template_id, :config_template_name, :hostgroup_id, :hostgroup_name, :environment_id, :environment_name
```
Comment (domcleal) [inline app/views/api/v2/template_combinations/base.json.rabl:3]: Lots of attributes for base?

File: app/views/api/v2/users/show.json.rabl [hunk 19]
```diff
@@ -1,8 +1,6 @@
 object @user
 
-attributes :id, :login, :firstname, :lastname, :mail, :admin, :auth_source_id, :auth_source_name, :last_login_on,
```
Comment (domcleal) [inline app/views/api/v2/users/show.json.rabl:3]: These attrs have been lost from show + index?

File: app/views/api/v2/users/show.json.rabl [hunk 20]
```diff
@@ -1,8 +1,6 @@
...
-           :domains_andor, :hostgroups_andor, :facts_andor, :filter_on_owner, :compute_resources_andor,
-           :created_at, :updated_at
+extends "api/v2/users/main"
 
 child :auth_source do
   extends "api/v2/auth_source_ldaps/show"
```
Comment (domcleal) [inline app/views/api/v2/users/show.json.rabl]: base?

**Comment by domcleal** [comment]
Broadly looks good, thanks.  Some things that came up a few times in my comments were:
1. base not always used for child nodes
2. some "create" templates use show, some use base - I think they should be show
3. some has_many relationships were just IDs rather than child nodes (unsure if you want to address this now though)

**Comment by isratrade** [comment]
@domcleal, I was just having this PR be adding the base, main, show templates and then clean up which attributes are in which file and DRY up things in another PR, but I will do it in this PR.

File: app/views/api/v2/architectures/show.json.rabl (hunk 1 above)
Comment (isratrade) [inline app/views/api/v2/architectures/show.json.rabl:3]: yes, has many attributes will be in PR #1062

**Comment by domcleal** [comment]
@isratrade ok, I think we should have a go at getting it right now, hopefully most are quick fixes anyway

File: app/views/api/v2/template_combinations/base.json.rabl (hunk 18 above)
Comment (isratrade) [inline app/views/api/v2/template_combinations/base.json.rabl:3]: @domcleal, yes I figured the combination is the only 'base' information.  There is no 'name' for example.

File: app/views/api/v2/reports/main.json.rabl (hunk 13 above)
Comment (isratrade) [inline app/views/api/v2/reports/main.json.rabl:5]: I added :host_id, :host_name to 'base', but it doesn't make "Show outdated diff"

File: app/views/api/v2/reports/show.json.rabl (hunk 14 above)
Comment (isratrade) [inline app/views/api/v2/reports/show.json.rabl:5]: @domcleal, this was removed from 'main'

File: app/views/api/v2/users/show.json.rabl (hunk 19 above)
Comment (isratrade) [inline app/views/api/v2/users/show.json.rabl:3]: attrs put back in 'main'

**Comment by isratrade** [comment]
@domcleal, re-committed as a separate commit to squash after review.

**Comment by isratrade** [comment]
@domcleal, it's much cleaner now.  Thanks for your **thorough** review. +1

**Comment by isratrade** [comment]
@mbacovsky, I moved `fullname` to operatingsystems/base so I don't need to move `major` and `minor`

**Comment by isratrade** [comment]
@mbacovsky, can you explain what you mean by your comment above
`What do you think about max line length in rabl templates or visually grouping the attributes? Some of the views are difficult to edit/check?`
Can you give an example?

**Comment by isratrade** [comment]
@mbacovsky, I added a new commit for audits/base and main.

**Comment by isratrade** [comment]
@mbacovsky, is there anything I read to change with api/plugins.  I think I may have squashed and fixed this already.

**Comment by domcleal** [comment]
Oh, maybe plugins doesn't need to change as the controller only supports index.

**Comment by mbacovsky** [comment]
:+1: 
I like the changes and I'm okay with the plugins.
@isratrade By max line length I meant, that the lines in the rabl are too long e.g. hosts/main.rabl. I'd prefer to have some style conventions to make it easier to read. This is example of longish rabl formatting that I find easier to read https://github.com/Katello/katello/blob/master/app/views/katello/api/v2/systems/show.json.rabl What do you think?

**Comment by domcleal** [comment]
I agree with @mbacovsky, this would be very useful for future reviews, though I'm going to merge this as-is for now.  Thanks for the extra pair of eyes!

Merged as 2e46934 for Foreman 1.4.0, thanks @isratrade.

**Comment by isratrade** [comment]
@mbacovsky, I agree with re-factoring in the future in the make the rabl files easier to read.

**Pull request closed without merging**